src/test/resources/regressions/examples/all_nearest_smaller_values.gobra

```diff
@@ -0,0 +1,218 @@
+requires acc(&left[j], 1/2) // full permission to `&left[j]`, read-only permissions to all other indices
+requires forall i int :: { left[i] } 1 <= i && i < j ==> 0 <= left[i] && left[i] < i &&
+                                   (0 < left[i] ==> s[left[i]] < s[i]) &&
+                                   forall k int :: { s[k] } left[i] < k && k <= i ==> s[i] <= s[k]
+ensures forall i int :: { s[i] } 1 <= i && i < len(s) ==> acc(&s[i], 1/4)
+ensures stack(res)
+ensures stackInv(res, s, j+1)
+ensures forall i int :: { left[i] } 1 <= i && i < len(left) ==> acc(&left[i], 1/2)
+ensures acc(&left[j], 1/2)
+ensures forall i int :: { left[i] } 1 <= i && i <= j ==> 0 <= left[i] && left[i] < i &&
+                                   (0 < left[i] ==> s[left[i]] < s[i]) &&
+                                   forall k int :: { s[k] } left[i] < k && k <= i ==> s[i] <= s[k]
+func codeOuterLoopBody(myStack *Stack, s []int, j int, left []int) (res *Stack) {
+    myStack = codeInnerLoop(myStack, s, j)
+
+    if (isEmpty(myStack)) {
+        left[j] = 0
+    } else {
+        left[j] = peek(myStack)
+    }
+
+    return pushWithInv(myStack, j, s, j+1)
+}
+
+requires 1 <= j && j < len(s)
+requires forall i int :: { s[i] } 1 <= i && i < len(s) ==> acc(&s[i], 1/8)
+requires stack(myStack)
+requires stackInv(myStack, s, j)
+ensures forall i int :: { s[i] } 1 <= i && i < len(s) ==> acc(&s[i], 1/8)
+ensures stack(res)
+ensures stackInv(res, s, j)
+ensures isEmpty(res) || s[peek(res)] < s[j]
+func codeInnerLoop(myStack *Stack, s []int, j int) (res *Stack) {
+    invariant 1 <= j && j < len(s)
+    invariant forall i int :: { s[i] } 1 <= i && i < len(s) ==> acc(&s[i], 1/8)
+    invariant stack(myStack)
+    invariant stackInv(myStack, s, j)
+    for !isEmpty(myStack) && s[peek(myStack)] >= s[j] {
+        unfold stack(myStack)
+        oldValue := myStack.value
+        myStack = myStack.next
+        convertStackInv(myStack, s, oldValue, j)
+    }
+    return myStack
+}
+
+requires 1 <= v && v < nz && nz <= len(s)
+requires stack(w)
+requires forall i int :: { s[i] } 1 <= i && i < len(s) ==> acc(&s[i], 1/8)
+requires stackInv(w, s, v)
+requires nz < len(s) ==> forall k int :: { s[k] } v < k && k <= nz ==> s[nz] <= s[k]
+ensures  nw != nil
+ensures  stack(nw)
+ensures  forall i int :: { s[i] } 1 <= i && i < len(s) ==> acc(&s[i], 1/8)
+ensures  unfolding stack(nw) in nw.next == w && nw.value == v
+ensures  stackInv(nw, s, nz)
+func pushWithInv(w *Stack, v int, ghost s []int, ghost nz int) (nw *Stack) {
+    // copy of push function. However, replacing it by a function call does not suffice to prove the postconditions
+    nw = &Stack{value: v, next: w}
+    fold stack(nw)
+}
+
+/* implemented algorithm (in func `code`):
+int i = 0
+stack s = emtpy
+while (i < N) {
+  h = stack.peek
+  while !s.isEmpty && s[h] >= s[i] {
+    stack.pop
+    h = stack.peek
+  }
+
+  if stack.isEmpty {
+    left[i] = 0
+  } else {
+    left[i] = h
+  }
+
+  stack.push(i)
+
+  i = i + 1
+}
+*/
```

Comment:
is this comment taken into account when measuring the LoC?

Reply:
I'll double check this